### PR TITLE
Add PreparedStatement support in SQLPipeline

### PR DESCRIPTION
**Pull request by lawben**
This PR adds the option to execute Prepared Statements. 

```cpp

auto prepared_statement_cache = std::make_shared<SQLQueryCache<SQLQueryPlan>>(5);

std::string prepare_statement = "PREPARE test FROM 'SELECT * FROM test WHERE a = ?'";
std::string execute_statement = "EXECUTE test(100)";

// Create SQLQueryPlan for SELECT * FROM test WHERE a = ?
SQLPipeline(prepare_statement, prepared_statement_cache).get_result_table();

// Bind ? from above statement with 100 and execute query
SQLPipeline(execute_statement, prepared_statement_cache).get_result_table();
```

Prepared statements are not defined globally, but per "session" (however exactly that is defined in Hyrise). So we will have a `prepared_statement_cache` in the console, and (later when the server is merged) one `prepared_statement_cache` per client connection. It is the task of the caller to create this cache.

This PR also contains refactorings to the `SQLPipeline` and `SQLPipelineStatement` constructors. We have many configurations for them so it made sense to create separate constructors for the different use cases. Having 4 parameters of which 3 are optional is very confusing and then suddenly the order is important because the other parameters need to be passed then too.

### WIP: 
- [x] still need to add to console
- [x] add some tests for more complicated queries
- [x] doesn't work for inserts. (the code is in PR #725. it probably makes sense to copy it into this PR because of tighter coupling of concerns)

### CODE REVIEW

File: src/lib/sql/sql_pipeline.cpp [hunk 1]
```diff
@@ -1,24 +1,58 @@
...
+  DebugAssert(prepared_statements != nullptr, "Cannot pass nullptr as explicit prepared statement cache.");
+}
 
-SQLPipeline::SQLPipeline(const std::string& sql, std::shared_ptr<opossum::TransactionContext> transaction_context,
-                         const std::shared_ptr<Optimizer>& optimizer)
-    : SQLPipeline(sql, transaction_context, UseMvcc::Yes, optimizer) {
+SQLPipeline::SQLPipeline(const std::string& sql, const std::shared_ptr<Optimizer>& optimizer,
+                         const PreparedStatementCache& prepared_statements,
+                         std::shared_ptr<opossum::TransactionContext> transaction_context)
+    : SQLPipeline(sql, transaction_context, UseMvcc::Yes, optimizer, prepared_statements) {
```
Comment (mrks) [inline src/lib/sql/sql_pipeline.cpp]: I really want designated initializers...

File: src/bin/console/console.cpp [hunk 2]
```diff
@@ -115,6 +115,8 @@ Console::Console()
   for (auto it = tpcc_generators.begin(); it != tpcc_generators.end(); ++it) {
     _tpcc_commands.push_back(it->first);
   }
+
+  _prepared_statement_cache = std::make_shared<SQLQueryCache<SQLQueryPlan>>(DefaultCacheCapacity);
```
Comment (mrks) [inline src/bin/console/console.cpp]: Not sure if we should call this a "cache" (where entries get evicted) or a store.

File: src/lib/sql/sql_pipeline_statement.cpp [hunk 3]
```diff
@@ -80,7 +119,15 @@ const std::shared_ptr<AbstractLQPNode>& SQLPipelineStatement::get_unoptimized_lo
...
 
-  const auto& parsed_sql = get_parsed_sql_statement();
+  auto parsed_sql = get_parsed_sql_statement();
+
+  const auto* statement = parsed_sql->getStatement(0);
+  if (const auto prepared_statement = dynamic_cast<const hsql::PrepareStatement*>(statement)) {
+    Assert(_prepared_statements, "Cannot prepare statement without prepared statement cache.");
+    parsed_sql = SQLPipelineStatement(prepared_statement->query).get_parsed_sql_statement();
+    _num_parameters = static_cast<uint16_t>(parsed_sql->parameters().size());
+  }
```
Comment (mrks) [inline src/lib/sql/sql_pipeline_statement.cpp]: Could you please add a comment to this block? If I understand it correctly, it is to "unwrap" the actual SQL statement from the prepare statement?

File: src/lib/sql/sql_pipeline_statement.cpp [hunk 4]
```diff
@@ -126,41 +173,64 @@ const std::shared_ptr<SQLQueryPlan>& SQLPipelineStatement::get_query_plan() {
...
+        Assert(_use_mvcc == UseMvcc::No, "Trying to use non-MVCC cached query with a transaction context.");
+      }
+
+      _query_plan->append_plan(plan.recreate());
+    } else if (const auto* execute_statement = dynamic_cast<const hsql::ExecuteStatement*>(statement)) {
+      // Handle query plan if we are executing a prepared statement
+      Assert(_prepared_statements, "Cannot execute statement without prepared statement cache.");
+      const auto plan = _prepared_statements->try_get(execute_statement->name);
+      Assert(plan, "Requested prepared statement does not exist!");
+
```
Comment (mrks) [inline src/lib/sql/sql_pipeline_statement.cpp]: Don't we have to do the MVCC check here as well?

File: src/lib/sql/sql_pipeline_statement.cpp [hunk 5]
```diff
@@ -126,41 +173,64 @@ const std::shared_ptr<SQLQueryPlan>& SQLPipelineStatement::get_query_plan() {
...
-  try {
-    _query_plan->add_tree_by_root(LQPTranslator{}.translate_node(lqp));
     if (_use_mvcc == UseMvcc::Yes) _query_plan->set_transaction_context(_transaction_context);
   } catch (const std::exception& exception) {
     throw std::runtime_error("Error while translating query plan:\n  " + std::string(exception.what()));
   }
 
-  const auto done = std::chrono::high_resolution_clock::now();
-  _compile_time_micros = std::chrono::duration_cast<std::chrono::microseconds>(done - started);
+  if (const auto* prepared_statement = dynamic_cast<const hsql::PrepareStatement*>(statement)) {
```
Comment (mrks) [inline src/lib/sql/sql_pipeline_statement.cpp]: The nesting in here is confusing. Why are some parts in the try-block and others not?

Comment (lawben) [reply src/lib/sql/sql_pipeline_statement.cpp]: Any suggestions on how to change this? The try catch block contains the 3 ways to get a QueryPlan and all of those can fail in some way that we want to catch. The rest is not relevant for the try-catch... Not sure how to change this to make it more intuitive. The only thing is can come up with is separate methods

Comment (mrks) [reply src/lib/sql/sql_pipeline_statement.cpp]: Mmh, ok. Let's leave it for now.

File: src/lib/sql/sql_pipeline_statement.cpp [hunk 6]
```diff
@@ -175,24 +175,31 @@ const std::shared_ptr<SQLQueryPlan>& SQLPipelineStatement::get_query_plan() {
...
       // Handle query plan if statement has been cached
       auto& plan = *cached_plan;
 
       DebugAssert(!plan.tree_roots().empty(), "QueryPlan retrieved from cache is empty.");
-      if (plan.tree_roots().front()->transaction_context_is_set()) {
-        Assert(_use_mvcc == UseMvcc::Yes, "Trying to use MVCC cached query without a transaction context.");
-      } else {
-        Assert(_use_mvcc == UseMvcc::No, "Trying to use non-MVCC cached query with a transaction context.");
-      }
+      check_same_mvcc_mode(plan);
```
Comment (mrks) [inline src/lib/sql/sql_pipeline_statement.cpp]: `assert_mvcc_mode_matches`?

Comment (mrks) [reply src/lib/sql/sql_pipeline_statement.cpp]: Also, should this be public? If not, we need an underscore.

Comment (lawben) [reply src/lib/sql/sql_pipeline_statement.cpp]: This is a simple lamdba that is defined within the function, so it's not public. But naming can be changed.

Comment (mrks) [reply src/lib/sql/sql_pipeline_statement.cpp]: oups, you're right

File: src/lib/sql/sql_pipeline.cpp (hunk 1 above)
Comment (lawben) [reply src/lib/sql/sql_pipeline.cpp]: ¯\_(ツ)_/¯

**Comment by mrks** [comment]
"Coverage — decreased by .4 compared to master" :(

**Comment by lawben** [comment]
@mrks something in a cache test still seems broken, don't know why/how/what but that might affect the coverage. where are you seeing that number?

**Comment by mrks** [comment]
That was from the last successful build.

**Comment by lawben** [comment]
@mrks: Coverage — increased by 1.0 compared to master and is now 89.4%... better?

**Comment by mrks** [comment]
:)

**Comment by lawben** [comment]
The CO build has been stuck on this line for at leat 40 minutes. Any idea why?

`[ 60%] Building ASM object src/lib/CMakeFiles/hyrise.dir/embed_llvm/jit_llvm_bundle.s.o`

In clangReleaseSanitizers only... maybe this should be looked into?

**Comment by mrks** [comment]
Let's keep an eye on that. Also paging @jfrohnhofen to let him know of this.

**Pull request merged**